### ISSUE TITLE
Audio bell does not work in integrated terminal

### ISSUE THREAD

**Issue by jhpratt** [microsoft/vscode#47711]
- VSCode Version: 1.22.1 (commit `950b8b0d37a9b7061b6f0d291837ccc4015f5ecd`)
- OS Version: Ubuntu 17.10

Steps to Reproduce:

1. Open an integrated terminal
2. Type `echo "\a"`

Assuming your OS is configured to allow TTY, you should hear a bell. In a normal terminal, this works fine. In code's, it does not.

Does this issue occur when all extensions are disabled?: Yes

**Comment by Tyriar** [comment]
Looks like there's an issue with the bell in the packages version of VS Code (not in the dev build):

`Uncaught (in promise) DOMException: Unable to decode audio data`

**Comment by Tyriar** [comment]
This may be related to Electron https://bugs.chromium.org/p/chromium/issues/detail?id=409402, should verify after we upgrade to Electron 2.

**Comment by Tyriar** [comment]
Not fixed in Electron 2?

**Comment by DrSensor** [comment]
I have the same problem with this in my machine:
```
Version: 1.31.1
Commit: 1b8e8302e405050205e69b59abb3559592bb9e60
Date: 2019-02-12T02:19:29.629Z
Electron: 3.1.2
Chrome: 66.0.3359.181
Node.js: 10.2.0
V8: 6.6.346.32
OS: Linux x64 4.19.13-1-MANJARO
```

**Comment by kvharish** [comment]
Hi, any idea when this will be fixed?

I am trying to add a bell sound at the end of my grunt task. It works fine in default terminal but not in the vscode integrated terminal.

`Version: 1.33.1 (1.33.1)
Commit: 51b0b28134d51361cf996d2f0a1c698247aeabd8 
Date: 2019-04-11T08:14:39.158Z
OS - macOS High Sierra (10.13.6)`

**Comment by jsejcksn** [comment]
I can also reproduce this issue in macOS with

```shell
echo -ne '\007'
```

**VS Code:** 1.34.0
**macOS:** 10.14.5

**Comment by matchu** [comment]
I'm seeing this too.

Even with `terminal.integrated.enableBell` enabled, running `printf "\a"` in the VS Code Integrated Terminal triggers no sound, despite the same command triggering a sound in my OS terminal.

**Comment by Tyriar** [comment]
This appears to work in the demo so I suspect it's actually a bug in vscode rather than an upstream issue.

**Comment by Tyriar** [comment]
Works fine out of sources, in the product build I see this:

```
ERR Unable to decode audio data
```

**Comment by Tyriar** [comment]
This is not related to CSP but because we strip out ffmpeg.

**Comment by Tyriar** [comment]
https://github.com/xtermjs/xterm.js/pull/2506

**Comment by Tyriar** [comment]
MP3s aren't decoded either, /cc @deepak1556

**Comment by matthewpoer** [comment]
Seeing this in VS Code Version: 1.41.1 on Catalina. 

Who's got the keys to the jeep? Not VS Code's terminal :(

**Comment by bycn** [comment]
Any updates on this?

**Comment by deepak1556** [comment]
@Tyriar sorry missed this one, `mp3` isn't supported with the ffmpeg library we ship with. Only the following audio codecs are supported `.opus`, `.webm`, `.flac`.

For testing,

The dev build uses electron release from https://github.com/electron/electron/releases which bundles ffmpeg library with proprietary codec support. For product build we ship with a minimal version of ffmpeg, so to test please extract the ffmpeg library from https://github.com/microsoft/vscode-electron-prebuilt/releases according to the platform and replace it under `vscode/.build/electron`

```
* macOS
vscode/.build/electron/Code - OSS.app/Contents/Frameworks/Electron Framework.framework/Versions/A/Libraries/libffmpeg.dylib

* linux
vscode/.build/electron/Code - OSS/libffmpeg.so

* windows
vscode/.build/electron/Code - OSS/libffmpeg.dll
```

**Comment by matthewpoer** [comment]
Hey @deepak1556 thanks for reaching out on this thread :)

I'm surprised to think that the system bell would be an mp3 file, but I like the idea that different build commands could cause this. I mean, I'd prefer it not be, but I think it's a good start on troubleshooting if it's true.

It looks like the repo @ `https://github.com/microsoft/vscode-electron-prebuilt` might not be publicly accessible? The URL `https://github.com/microsoft/vscode-electron-prebuilt/releases` is a dead link for me.

Using that idea, though, I grabbed `ffmpeg-4.2.2.zip` from https://evermeet.cx/ffmpeg/ (linked to from [ffmpeg's official download page](https://www.ffmpeg.org/download.html#build-mac)), but it looks like this is a full-on executable whereas VS Code just has a library-specific build.

> file libffmpeg.dylib
> libffmpeg.dylib: Mach-O 64-bit dynamically linked shared library x86_64

I'm not familiar with building ffmpeg, but would it be possible for you to extract the build command that the `vscode-electron-prebuilt` repo. is using to generate the library? Or if it's easier, upload the pre-built binary?

For reference, I found my local path to the ffmpeg library for VS Code as:
```
/Applications/Visual Studio Code.app/Contents/Frameworks/Electron Framework.framework/Versions/A/Libraries/libffmpeg.dylib
```

**Comment by deepak1556** [comment]
> The URL https://github.com/microsoft/vscode-electron-prebuilt/releases is a dead link for me.

The repo is private :)

> but it looks like this is a full-on executable whereas VS Code just has a library-specific build.

Yeah VSCode or rather electron just relies on the shared library, which makes it easier to switch between libraries that have full codec support.

For your case, test with the library from https://github.com/electron/electron/releases/download/v7.1.11/ffmpeg-v7.1.11-darwin-x64.zip

**Comment by matthewpoer** [comment]
Cool, grabbed the library, swapped it in and `printf '\a'` now makes a weird little noise. It's kind of soft and I think you could probably miss it if you weren't expecting it. Definitely not the system bell I'm accustomed to (i.e. `printf '\a'` in a Terminal window is a different sound). Do you know what would account for this?

**Comment by deepak1556** [comment]
> Definitely not the system bell I'm accustomed to (i.e. printf '\a' in a Terminal window is a different sound)

That would be from https://github.com/xtermjs/xterm.js/blob/master/src/common/services/OptionsService.ts#L15 

@Tyriar can talk for why it was chosen, as I don't have context on it.

But the issue here is the codec supported by ffmpeg we ship with, which should be fixed once we adopt the right one.

**Comment by Tyriar** [comment]
@deepak1556 I just found a public domain bell-like sound. xterm.js is a web app and doesn't have access to some system bell sound.

**Comment by KestutisMa** [comment]
Using `speaker-test` in vscode terminal (Arch Linux) also fails to emit audio.
Outside vscode it is working fine.
```
speaker-test -t waw -c 6
ALSA lib pcm_dmix.c:1089:(snd_pcm_dmix_open) unable to open slave
Playback open error: -2,No such file or directory

**Comment by fletch29** [comment]
Electron appears to offer a way to invoke the default system sound:
shell.beep()

https://www.electronjs.org/docs/api/shell#shellbeep

**Comment by Tyriar** [comment]
@fletch29 nice find, I created an upstream issue that blocks supporting it: https://github.com/xtermjs/xterm.js/issues/3014

**Comment by denishowew** [comment]
MacOS work-around:
`osascript -e "beep beep beep"`

**Comment by matthewpoer** [comment]
> MacOS work-around:
> `osascript -e "beep beep beep"`

Yes! but let's go deeper and create a bash alias that takes over `beep` right in `~/.profile`

```
alias beep="osascript -e 'beep'"
```

**Comment by mrt123** [comment]
Same issue on : 
Version: 1.52.1
Commit: ea3859d4ba2f3e577a159bc91e3074c5d85c0523
Date: 2020-12-16T16:30:02.420Z
Electron: 9.3.5
Chrome: 83.0.4103.122
Node.js: 12.14.1
V8: 8.3.110.13-electron.0
OS: Darwin x64 18.7.0

Tested via nodejs with : console.log("\x07")
workaround : 
```
const ringSpeaker = () => {
    const { exec } = require("child_process");
    exec('osascript -e "beep beep beep"'); 
};
```
Wish it worked in a loop, so that function could accept nTimes argument for number of beeps.

**Comment by bobmagicii** [comment]
here is a +1 from 2021

**Comment by Mellbourn** [comment]
It's nice that there are workarounds like `osascript`, but many existing shell scripts expect the actual bell (`tput bel`) to work and don't allow any customization to replace it with anything else.

**Comment by Tyriar** [comment]
I know it's not the sound, but a visual bell is coming in the upcoming release which is driven by the same setting: https://github.com/microsoft/vscode/issues/121939

![image](https://user-images.githubusercontent.com/2193314/115723847-3bfd9200-a335-11eb-907a-c14eb766f643.png)

**Comment by tjx666** [comment]
Should also be shown in none-tabs mode?
![image](https://user-images.githubusercontent.com/41773861/116281752-988d0100-a7bc-11eb-9b78-16d21524738f.png)

**Comment by Tyriar** [comment]
@tjx666 I think it will look bad if we change the title in the dropdown as other titles will shift around when the bell triggers. We might end up adding a visual bell mode later on when this issue gets tackled that will flash the border around the terminal or something.

**Comment by Tyriar** [comment]
Moved the support for the bell in dropdown to https://github.com/microsoft/vscode/issues/128859

**Comment by Mellbourn** [comment]
I think an *audible* bell would be preferable

**Comment by cliffordfajardo** [comment]
Just wanted to add a data point about this commands usefulness:

I typically run commands that take 1-2 minutes to complete in the terminal. _Ex: building or compiling a program_
In iTerm or terminal app, I tend to append the `beep` command at the end of my program to let me know when its done.

```
npm run build && beep
```

## Work around for now
This aliases bellow would be added to your `.bashrc`, `.zshrc` files etc
```sh
alias beep="osascript -e 'beep'"✅  # ring 1 time

# Below are 2  examples of customizeable version of beep

alias beep='function donebeep(){ for i in {1..10}; do osascript -e beep; done };donebeep' # run 10 times
alias donebeep='function donebeep(){ for i in {1..$1}; do osascript -e beep; done };donebeep' # run N time
```

Usage:

```sh
npm run build && donebeep 10
```

**Comment by jivanpal** [comment]
On macOS, I use the `say` command as a workaround.

**Comment by Abesyuffcjk** [comment]
#142400

**Comment by Blealtan** [comment]
+1 from 2022.
My use case is in Remote mode. I work through Remote SSH/WSL, and expect `printf \\a` to ring the bell when my build/test script finished. Many terminal emulators, e.g. Windows Terminal, work well in such cases.

**Comment by sunshaoce** [comment]
When the compilation time is very long, if there is a ringtone, I don't have to stare at the screen all the time.

**Comment by tjx666** [comment]
@Tyriar @meganrogge seems this can be implemented because of Shell integration?

**Comment by Tyriar** [comment]
@tjx666 it was never lacking shell integration, it's an issue with Electron sound. We seem to have worked around those since audio cues are now supported, this just needs someone to investigate and create a sound (which I've requested internally).

**Comment by FreTimmerman** [comment]
FYI: it's not just the sound
in a normal terminal on macOS, the command `printf "\a"` will (besides beeping) make the terminal icon jump up in the dock, and add a notification bubble to it.

the built-in vscode terminal does none of those

**Comment by Bauxitedev** [comment]
On Windows, if you're remoting into WSL, you can use this as a workaround:
```
powershell.exe '[console]::beep(261.6,700)'
```

This will play a short beep sound.

**Comment by tjx666** [comment]
I see last release of vscode fix some accessbility issues, the audio bell should also be a accessbility issue. I think this is a welcome feature but not hard to implement. Any plan to implement it?

**Comment by Tyriar** [comment]
@tjx666 I agree, this was mislabeled. I was waiting on sounds from another MS team before continuing which seems to have stalled so I pinged them again.

**Comment by nnWhisperer** [comment]
Thank you. How long would it take to have it in the main Vscode version(not insiders) ?

**Comment by meganrogge** [comment]
we will release this version in a few weeks @nnWhisperer

**Comment by jivanpal** [comment]
Is this supposed to be resolved as of 1.72.2 (commit d045a5eda657f4d7b676dedbfa7aab8207f8a075)? If so, it is not working for me on macOS 10.15.7; running `printf '\a'` in the VSCode terminal does nothing.

**Comment by meganrogge** [comment]
You need to opt in w the setting for it to work (it's off by default)

**Comment by 0xDECAFC0FFEE** [comment]
The audio bell works on my computer!
Thanks for your work - this has been a huge pain point for my workflow for years.

**Comment by matthewpoer** [comment]
Found myself on `Version: 1.72.2 ` and can confirm that enabling the setting does in fact give you a bell sound :D 

Enable with `Cmd + Shift + P`, search/select `Preferences: Open Settings (UI)`, search for "bell" and check on the "Terminal > Integrated: Enable Bell" toggle.

![screenshot of settings](https://user-images.githubusercontent.com/727257/198130377-b145e363-8e97-4005-9764-3216a0a487b1.png)

Can also use

```json
"terminal.integrated.enableBell": true
```

Moving forward `printf '\a'` will give you a chime.

**Comment by jivanpal** [comment]
Thanks all, it's working now. However, it doesn't use the alert sound I have configured on my Mac ("Glass" ding in System Sounds), and the "bloop" sound that VSCode uses is much too quiet; I have to turn my volume up to 100% for it to be at the same level as the Glass sound at 25%. Is there a way to adjust the volume level and/or the sound file that is played?

**Comment by FreTimmerman** [comment]
i can confirm `printf '\a'` will give a chime. BUT:

it is NOT the system chime, NOR does it bump the VScode icon in the dock, NOR does it add a notification icon to VScode in the dock. (terminal does all this, see icon attached)
<img width="64" alt="image" src="https://user-images.githubusercontent.com/571115/198252258-60857c0a-df02-4f6d-9789-ba6ed76f9059.png">

**Comment by jivanpal** [comment]
@denishowew For getting notified when a command/script finishes, I've been using `say` for some time, as suggested by others here. However, the chime/alert sound plays in many other situations, too, such as when a tab-autocomplete or history reverse-search (`^R`) fails.
